src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe_opt.cpp

```diff
@@ -0,0 +1,1126 @@
+            args.insert({DNNL_ARG_ATTR_SCALES | DNNL_ARG_WEIGHTS, scale});
+        }
+        if (zp) {
+            args.insert({DNNL_ARG_ATTR_ZERO_POINTS | DNNL_ARG_WEIGHTS, zp});
+        }
+        if (bin_mem) {
+            args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(bin_post_id) | DNNL_ARG_SRC_1, bin_mem});
+        }
+        m_prim.execute(stream, args);
+    }
+};
+
+class MOEOptSoftMaxTopK : public KernelGenerator {
+public:
+    MOEOptSoftMaxTopK() : KernelGenerator("moe_opt", "softmax_topk") {}
+
+protected:
+    [[nodiscard]] JitConstants get_jit_constants(const RuntimeParams& params) const override {
+        auto jit = KernelGenerator::get_jit_constants(params);
+        auto desc = params.typed_desc<moe_fused_compressed>();
+        jit.make("SOFTMAX_TOPK_ENABLE", 1);
+        jit.make("TOP_K", desc->_config.top_k);
+        jit.make("VALUE_NUM", desc->_config.num_expert);
+        jit.make("MOE_TYPE", params.get_input_layout(0).data_type == ov::element::f16 ? "half" : "float");
+        jit.make("MOE_TYPE_SIZE", params.get_input_layout(0).data_type == ov::element::f16 ? 2 : 4);
+        return jit;
+    }
+
+    [[nodiscard]] Arguments get_arguments_desc(const RuntimeParams& params) const override {
+        Arguments args;
+
+        return args;
+    }
+
+    [[nodiscard]] DispatchDataFunc get_dispatch_data_func() const override {
+        return DispatchDataFunc{nullptr};
+    }
+};
+
+class MOEOptGather : public KernelGenerator {
+public:
+    MOEOptGather() : KernelGenerator("moe_opt", "gather") {}
+
+protected:
+    [[nodiscard]] JitConstants get_jit_constants(const RuntimeParams& params) const override {
+        auto jit = KernelGenerator::get_jit_constants(params);
+        auto desc = params.typed_desc<moe_fused_compressed>();
+        auto& engine = params.prog->get_engine();
+        const auto& info = engine.get_device_info();
+        jit.make("GATHER_ENABLE", 1);
+        jit.make("HIDDEN_SIZE", desc->_config.hidden_size);
+        jit.make("MOE_TYPE", params.get_input_layout(0).data_type == ov::element::f16 ? "half" : "float");
+        jit.make("MOE_TYPE_SIZE", params.get_input_layout(0).data_type == ov::element::f16 ? 2 : 4);
+        jit.make("SUBGROUP_SIZE", info.arch >= gpu_arch::xe2 ? 32 : 16);
+        return jit;
+    }
+
+    [[nodiscard]] Arguments get_arguments_desc(const RuntimeParams& params) const override {
+        Arguments args;
+
+        return args;
+    }
+
+    [[nodiscard]] DispatchDataFunc get_dispatch_data_func() const override {
+        return DispatchDataFunc{nullptr};
+    }
+};
+
+class MOEOptScatter : public KernelGenerator {
+public:
+    MOEOptScatter() : KernelGenerator("moe_opt", "index_add") {}
+
+protected:
+    [[nodiscard]] JitConstants get_jit_constants(const RuntimeParams& params) const override {
+        auto jit = KernelGenerator::get_jit_constants(params);
+        auto desc = params.typed_desc<moe_fused_compressed>();
+        jit.make("SCATTER_ENABLE", 1);
```

Comment:
BTW why not having the sub kernels in separate file?

Reply:
Put it into a separated file is ok, but it will lead to too many cl files which should belong to one primitive.